spock-core/src/main/java/org/spockframework/mock/runtime/ProxyBasedMockFactory.java

```diff
@@ -74,9 +90,47 @@ private Object createDynamicProxyMock(Class<?> mockType, List<Class<?>> addition
   }
 
   // inner class to defer class loading
-  private static class CglibMockFactory {
+  private static class ByteBuddyMockFactory {
+
+    private static final ByteBuddyClassCache CACHE = new ByteBuddyClassCache(true);
+
     static Object createMock(Class<?> type, List<Class<?>> additionalInterfaces, @Nullable List<Object> constructorArgs,
         IProxyBasedMockInterceptor interceptor, ClassLoader classLoader, boolean useObjenesis) {
+      Class<?> enhancedType = CACHE.find(classLoader, type, additionalInterfaces);
+
+      if (enhancedType == null) {
+        synchronized (type) {
```

Comment:
The implicit cglib cache locks on the class object already today.

You are right that this can cause problems if the classobject is already locked by another thread. Using a single monitor for this would however not allow creating mocks for different classes from different threads. The question is, what limitation is preferred.